File: src/lib/rag/micro-graph-extractor.ts
import { useApiStore } from '../../store/api-store';
import { useSettingsStore } from '../../store/settings-store';
import { createLlmClient, ExtendedModelConfig } from '../llm/factory';
import { graphStore } from './graph-store';
import { getDefaultKgPrompt, getKgFreeModePrompt, getKgDomainAutoPrompt } from './defaults';
import { db } from '../db';
import { SearchResult } from './vector-store';

export interface MicroGraphResult {
  nodes: Array<{ name: string; type: string; metadata?: any }>;
  edges: Array<{ source: string; target: string; relation: string; weight?: number }>;
  context: string;
  sourceChunkIds: string[];
  query: string;
  extractedAt: number;
}

export class MicroGraphExtractor {
  /**
   * 基于召回文本动态提取微型图谱
   */
  async extract(
    topKResults: SearchResult[],
    query: string,
    sessionId: string,
    options?: {
      timeout?: number;      // 默认 5000ms
      maxChars?: number;     // 总输入截断 (默认 6000)
      onProgress?: (subStage: string) => void;
    }
  ): Promise<MicroGraphResult | null> {
    const timeout = options?.timeout || 5000;
    const maxChars = options?.maxChars || 6000;

    // 1. Cache Check
    const cacheKey = this.generateCacheKey(query, topKResults);
    const cached = await this.getFromCache(cacheKey);
    if (cached) {
      console.log('[MicroGraph] Cache HIT for key:', cacheKey.substring(0, 8));
      return cached;
    }

    try {
      // 2. Prepare Payload
      const textToAnalyze = this.prepareInputText(topKResults, maxChars);
      if (!textToAnalyze) return null;

      // 3. Extraction with Timeout
      options?.onProgress?.('KG_JIT');
      const extractionPromise = this.performExtraction(textToAnalyze, query, options?.onProgress);
      const timeoutPromise = new Promise<null>((resolve) => 
        setTimeout(() => resolve(null), timeout)
      );

      const result = await Promise.race([extractionPromise, timeoutPromise]);

      if (result) {
        const fullResult: MicroGraphResult = {
          ...result,
          sourceChunkIds: topKResults.map(r => r.id),
          query,
          extractedAt: Date.now()
        };
        // 4. Save to cache
        await this.saveToCache(cacheKey, fullResult);
        // 5. Background merge (Don't await)
        this.backgroundMerge(fullResult, sessionId);
        
        return fullResult;
      }

      console.warn('[MicroGraph] Extraction TIMEOUT or FAILED');
      return null;
    } catch (e) {
      console.error('[MicroGraph] Critical error during extraction:', e);
      return null;
    }
  }

  private generateCacheKey(query: string, results: SearchResult[]): string {
    const ids = results.map(r => r.id).sort().join(',');
    return `${query}:${ids}`;
  }

  private async getFromCache(key: string): Promise<MicroGraphResult | null> {
    try {
      // Opportunistic cleanup: delete expired entries on cache read
      await this.cleanupExpiredCache();

      const res = await db.execute(
        'SELECT result_json FROM kg_jit_cache WHERE cache_key = ? AND expires_at > ?',
        [key, Date.now()]
      );
      if (res.rows && res.rows.length > 0) {
        return JSON.parse((res.rows as any)[0].result_json);
      }
    } catch (e) {
      console.warn('[MicroGraph] Cache read failed:', e);
    }
    return null;
  }

  private async saveToCache(key: string, result: MicroGraphResult): Promise<void> {
    try {
      const settings = useSettingsStore.getState();
      const ttl = (settings.globalRagConfig.jitCacheTTL || 3600) * 1000;
      const expiresAt = Date.now() + ttl;

      await db.execute(
        'INSERT OR REPLACE INTO kg_jit_cache (cache_key, query_hash, chunk_ids_hash, result_json, created_at, expires_at) VALUES (?, ?, ?, ?, ?, ?)',
        [key, result.query, result.sourceChunkIds.join(','), JSON.stringify(result), Date.now(), expiresAt]
      );
    } catch (e) {
      console.warn('[MicroGraph] Cache save failed:', e);
    }
  }

  /**
   * 清理过期的 JIT 缓存条目
   */
  private async cleanupExpiredCache(): Promise<void> {
    try {
      await db.execute(
        'DELETE FROM kg_jit_cache WHERE expires_at <= ?',
        [Date.now()]
      );
    } catch (e) {
      // Silently ignore — cleanup is opportunistic
    }
  }

  private prepareInputText(results: SearchResult[], maxChars: number): string {
    let totalText = '';
    for (const res of results) {
      const remaining = maxChars - totalText.length;
      if (remaining <= 0) break;
      totalText += res.content.substring(0, remaining) + '\n---\n';
    }
    return totalText.trim();
  }

  /**
   * 提取 LLM 响应中的 JSON（三层 fallback，与 graph-extractor.ts 保持一致）
   */
  private extractJSONFromContent(content: string): any | null {
    let jsonString = content.trim();

    // Layer 1: ```json ... ```
    const jsonBlockMatch = jsonString.match(/```json\s*([\s\S]*?)\s*```/i);
    if (jsonBlockMatch) {
      jsonString = jsonBlockMatch[1].trim();
    } else {
      // Layer 2: generic code block that starts with {
      const genericMatch = jsonString.match(/```\s*([\s\S]*?)\s*```/);
      if (genericMatch) {
        const potential = genericMatch[1].trim();
        if (potential.startsWith('{')) {
          jsonString = potential;
        }
      } else {
        // Layer 3: outermost { ... }
        const first = jsonString.indexOf('{');
        const last = jsonString.lastIndexOf('}');
        if (first !== -1 && last !== -1 && last > first) {
          jsonString = jsonString.substring(first, last + 1);
        }
      }
    }

    try {
      return JSON.parse(jsonString);
    } catch (e) {
      console.warn('[MicroGraph] JSON parse failed:', e);
      console.log('[MicroGraph] Raw output preview:', content.slice(0, 200) + '...');
      return null;
    }
  }

  /**
   * 组装 JIT 专用 system prompt（复用 graph-extractor 的策略）
   */
  private buildJitPrompt(): string {
    const settingsState = useSettingsStore.getState();
    const config = settingsState.globalRagConfig;
    const customPrompt = config.kgExtractionPrompt;
    const entityTypes = config.kgEntityTypes || [];

    // 1. Base prompt
    let basePrompt = '';
    const isFreeMode = config.kgFreeMode || !entityTypes || entityTypes.length === 0;

    if (customPrompt && customPrompt.trim().length > 0) {
      basePrompt = customPrompt;
      if (basePrompt.includes('{entityTypes}')) {
        basePrompt = basePrompt.replace('{entityTypes}', entityTypes.length > 0 ? entityTypes.join(', ') : 'any meaningful types');
      }
    } else {
      basePrompt = isFreeMode ? getKgFreeModePrompt() : getDefaultKgPrompt();
      if (!isFreeMode && basePrompt.includes('{entityTypes}')) {
        basePrompt = basePrompt.replace('{entityTypes}', entityTypes.join(', '));
      }
    }

    // 2. Domain instruction (unified: kgDomainAuto OR kgDomainHint==='auto')
    if (config.kgDomainHint === 'auto' || config.kgDomainAuto) {
      basePrompt += '\n\n' + getKgDomainAutoPrompt();
    } else if (config.kgDomainHint) {
      basePrompt += `\n\nFocus on the context of: ${config.kgDomainHint}`;
    }

    return basePrompt;
  }

  private async performExtraction(text: string, query: string, onProgress?: (s: string) => void): Promise<any> {
    try {
      const settingsState = useSettingsStore.getState();
      const apiState = useApiStore.getState();

      onProgress?.('KG_JIT_PARSE');

      // Model Resolution
      const modelId = settingsState.globalRagConfig.kgExtractionModel || settingsState.defaultSummaryModel;
      let provider = apiState.providers.find(p => p.enabled && p.models.some(m => m.id === modelId || m.uuid === modelId));

      if (!provider) {
        provider = apiState.providers.find(p => p.enabled && p.models.length > 0);
      }

      if (!provider || !modelId) {
        console.warn('[MicroGraph] No model available for JIT KG extraction');
        return null;
      }

      const model = provider.models.find(m => m.id === modelId || m.uuid === modelId) || provider.models[0];
      const client = createLlmClient({
        ...model,
        provider: provider.type,
        apiKey: provider.apiKey,
        baseUrl: provider.baseUrl,
        vertexProject: provider.vertexProject,
        vertexLocation: provider.vertexLocation,
        vertexKeyJson: provider.vertexKeyJson,
      });

      // Build system prompt using unified strategy
      const systemPrompt = this.buildJitPrompt();

      const response = await client.chatCompletion([
        { role: 'system', content: systemPrompt },
        { role: 'user', content: `Query Context: "${query}"\n\nAnalyze the following text fragments and build a micro-graph of related entities:\n\n${text}` }
      ]);

      const content = response.content;
      if (!content) return null;

      // Three-layer JSON extraction (consistent with graph-extractor.ts)
      const result = this.extractJSONFromContent(content);
      if (!result || !result.nodes || !result.edges) {
        console.warn('[MicroGraph] Failed to extract valid JSON with nodes/edges from LLM response');
        return null;
      }

      // Build context string from edges
      const context = result.edges?.map((e: any) => `- ${e.source} --[${e.relation}]--> ${e.target}`).join('\n') || '';

      return { ...result, context };
    } catch (e) {
      console.error('[MicroGraph] performExtraction error:', e);
      return null;
    }
  }

  private async backgroundMerge(result: MicroGraphResult, sessionId: string): Promise<void> {
    try {
      console.log(`[MicroGraph] Background merging ${result.nodes.length} nodes to Global Graph...`);
      const nameToIdMap = new Map<string, string>();

      // Transactional merge
      await db.execute('BEGIN TRANSACTION');
      
      for (const node of result.nodes) {
        const id = await graphStore.upsertNode(node.name, node.type, node.metadata, { sessionId }, 'jit');
        nameToIdMap.set(node.name, id);
      }

      for (const edge of result.edges) {
        const sourceId = nameToIdMap.get(edge.source);
        const targetId = nameToIdMap.get(edge.target);
        if (sourceId && targetId) {
          await graphStore.createEdge(sourceId, targetId, edge.relation, undefined, edge.weight || 1.0, { sessionId }, 'jit');
        }
      }

      await db.execute('COMMIT');
      console.log('[MicroGraph] Background merge SUCCESS');
    } catch (e) {
      await db.execute('ROLLBACK');
      console.warn('[MicroGraph] Background merge failed:', e);
    }
  }
}

export const microGraphExtractor = new MicroGraphExtractor();

File: native-ui/app/src/main/java/com/promenar/nexara/data/rag/MicroGraphExtractor.kt
package com.promenar.nexara.data.rag

import com.promenar.nexara.data.local.db.dao.KgJitCacheDao
import com.promenar.nexara.data.local.db.entity.KgJitCacheEntity
import com.promenar.nexara.data.remote.protocol.LlmProtocol
import com.promenar.nexara.data.remote.protocol.PromptRequest
import com.promenar.nexara.data.remote.protocol.ProtocolMessage
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class MicroGraphExtractor(
    private val protocol: LlmProtocol,
    private val graphStore: GraphStore,
    private val jitCacheDao: KgJitCacheDao,
    private val modelId: String? = null,
    private val systemPrompt: String = GraphExtractor.DEFAULT_KG_PROMPT,
    private val cacheTtlSeconds: Long = 3600
) {
    suspend fun extract(
        topKResults: List<SearchResult>,
        query: String,
        sessionId: String,
        timeout: Long = 5000,
        maxChars: Int = 6000
    ): MicroGraphResult? {
        val cacheKey = generateCacheKey(query, topKResults)

        val cached = getFromCache(cacheKey)
        if (cached != null) return cached

        val textToAnalyze = prepareInputText(topKResults, maxChars)
        if (textToAnalyze.isBlank()) return null

        val extractionResult = performExtraction(textToAnalyze, query)
        if (extractionResult == null) return null

        val fullResult = MicroGraphResult(
            nodes = extractionResult.nodes,
            edges = extractionResult.edges,
            context = extractionResult.edges.joinToString("\n") { edge ->
                "- ${edge.source} --[${edge.relation}]--> ${edge.target}"
            },
            sourceChunkIds = topKResults.map { it.id },
            query = query,
            extractedAt = System.currentTimeMillis()
        )

        saveToCache(cacheKey, fullResult)
        backgroundMerge(fullResult, sessionId)

        return fullResult
    }

    private suspend fun getFromCache(key: String): MicroGraphResult? {
        try {
            jitCacheDao.deleteExpired(System.currentTimeMillis())
            val cached = jitCacheDao.getByKey(key) ?: return null
            return try {
                Json.decodeFromString<MicroGraphResult>(cached.resultJson)
            } catch (e: Exception) {
                null
            }
        } catch (e: Exception) {
            return null
        }
    }

    private suspend fun saveToCache(key: String, result: MicroGraphResult) {
        try {
            val now = System.currentTimeMillis()
            val expiresAt = now + cacheTtlSeconds * 1000
            val json = Json.encodeToString(MicroGraphResult.serializer(), result)
            jitCacheDao.insert(
                KgJitCacheEntity(
                    cacheKey = key,
                    queryHash = result.query,
                    chunkIdsHash = result.sourceChunkIds.joinToString(","),
                    resultJson = json,
                    createdAt = now,
                    expiresAt = expiresAt
                )
            )
        } catch (e: Exception) {
            // Cache save failure is non-critical
        }
    }

    private suspend fun performExtraction(text: String, query: String): ExtractionResult? {
        return try {
            val effectiveModelId = modelId ?: "default"
            val request = PromptRequest(
                messages = listOf(
                    ProtocolMessage(role = "system", content = systemPrompt),
                    ProtocolMessage(
                        role = "user",
                        content = """Query Context: "$query"

Analyze the following text fragments and build a micro-graph of related entities:

$text"""
                    )
                ),
                model = effectiveModelId,
                temperature = 0.0,
                stream = false
            )

            val response = protocol.sendPromptSync(request)
            val content = response.content
            if (content.isBlank()) return null

            parseJsonFromContent(content)
        } catch (e: Exception) {
            null
        }
    }

    private fun parseJsonFromContent(content: String): ExtractionResult? {
        var jsonString = content.trim()

        val jsonBlockRegex = Regex("""```json\s*([\s\S]*?)\s*```""", RegexOption.IGNORE_CASE)
        val genericBlockRegex = Regex("""```\s*([\s\S]*?)\s*```""")

        val jsonMatch = jsonBlockRegex.find(jsonString)
        val genericMatch = genericBlockRegex.find(jsonString)

        jsonString = when {
            jsonMatch != null -> jsonMatch.groupValues[1].trim()
            genericMatch != null -> {
                val potential = genericMatch.groupValues[1].trim()
                if (potential.startsWith("{")) potential else jsonString
            }
            else -> {
                val first = jsonString.indexOf('{')
                val last = jsonString.lastIndexOf('}')
                if (first != -1 && last != -1 && last > first) {
                    jsonString.substring(first, last + 1)
                } else jsonString
            }
        }

        return try {
            val jsonEl = kotlinx.serialization.json.Json.parseToJsonElement(jsonString).jsonObject

            val nodes = jsonEl["nodes"]?.jsonArray?.map { nodeEl ->
                val obj = nodeEl.jsonObject
                ExtractedNode(
                    name = obj["name"]?.jsonPrimitive?.content ?: "",
                    type = obj["type"]?.jsonPrimitive?.content ?: "concept",
                    metadata = obj["metadata"]?.toString()
                )
            } ?: emptyList()

            val edges = jsonEl["edges"]?.jsonArray?.map { edgeEl ->
                val obj = edgeEl.jsonObject
                ExtractedEdge(
                    source = obj["source"]?.jsonPrimitive?.content ?: "",
                    target = obj["target"]?.jsonPrimitive?.content ?: "",
                    relation = obj["relation"]?.jsonPrimitive?.content ?: "",
                    weight = obj["weight"]?.jsonPrimitive?.content?.toDoubleOrNull() ?: 1.0
                )
            } ?: emptyList()

            if (nodes.isEmpty() && edges.isEmpty()) return null
            ExtractionResult(nodes = nodes, edges = edges)
        } catch (e: Exception) {
            null
        }
    }

    private suspend fun backgroundMerge(result: MicroGraphResult, sessionId: String) {
        try {
            val nameToIdMap = mutableMapOf<String, String>()

            for (node in result.nodes) {
                val id = graphStore.upsertNode(
                    name = node.name,
                    type = node.type,
                    metadata = node.metadata,
                    scope = KgScope(sessionId = sessionId),
                    sourceType = "jit"
                )
                nameToIdMap[node.name] = id
            }

            for (edge in result.edges) {
                val sourceId = nameToIdMap[edge.source]
                val targetId = nameToIdMap[edge.target]
                if (sourceId != null && targetId != null) {
                    graphStore.createEdge(
                        sourceId = sourceId,
                        targetId = targetId,
                        relation = edge.relation,
                        docId = null,
                        weight = edge.weight,
                        scope = KgScope(sessionId = sessionId),
                        sourceType = "jit"
                    )
                }
            }
        } catch (e: Exception) {
            // Background merge failure is non-critical
        }
    }

    private fun generateCacheKey(query: String, results: List<SearchResult>): String {
        val ids = results.map { it.id }.sorted().joinToString(",")
        return "$query:$ids"
    }

    private fun prepareInputText(results: List<SearchResult>, maxChars: Int): String {
        val sb = StringBuilder()
        for (res in results) {
            val remaining = maxChars - sb.length
            if (remaining <= 0) break
            sb.append(res.content.take(remaining))
            sb.append("\n---\n")
        }
        return sb.toString().trim()
    }
}
